timemachine/cpp/src/local_md_potentials.cu

```diff
@@ -0,0 +1,190 @@
+#include "constants.hpp"
+#include "gpu_utils.cuh"
+#include "kernel_utils.cuh"
+#include "kernels/k_flat_bottom_bond.cuh"
+#include "kernels/k_indices.cuh"
+#include "kernels/k_local_md.cuh"
+#include "local_md_potentials.hpp"
+#include "math_utils.cuh"
+#include <cub/cub.cuh>
+#include <random>
+#include <vector>
+
+namespace timemachine {
+
+// Struct to as a CUB <= operation
+struct LessThan {
+    int compare;
+    CUB_RUNTIME_FUNCTION __device__ __forceinline__ explicit LessThan(int compare) : compare(compare) {}
+    CUB_RUNTIME_FUNCTION __device__ __forceinline__ bool operator()(const int &a) const { return (a < compare); }
+};
+
+LocalMDPotentials::LocalMDPotentials(const int N, const std::vector<std::shared_ptr<BoundPotential>> bps)
+    : N_(N), temp_storage_bytes_(0), all_potentials_(bps), restraints_(N_ * 2), bond_params_(N_ * 3),
+      probability_buffer_(round_up_even(N_)), d_free_idxs_(N_), d_row_idxs_(N_), d_col_idxs_(N_), p_num_selected_(1),
+      num_selected_buffer_(1) {
+
+    std::vector<std::shared_ptr<BoundPotential>> nonbonded_pots;
+    get_nonbonded_all_pair_potentials(bps, nonbonded_pots);
+
+    if (nonbonded_pots.size() > 1) {
+        throw std::runtime_error("found multiple NonbondedAllPairs potentials");
+    }
+    if (nonbonded_pots.size() != 1) {
+        throw std::runtime_error("unable to find a NonbondedAllPairs potential");
+    }
+
+    // Only used to reference shared_ptr to potential and for Nonbonded parameters
+    // modifications to the BoundPotential has no impact
+    nonbonded_bp_ = nonbonded_pots[0];
+
+    // Ensure that we allocate enough space for all potential bonds
+    std::vector<int> default_bonds(2 * N_);
+    for (int i = 0; i < N_; i++) {
+        default_bonds[i * 2 + 0] = 0;
+        default_bonds[i * 2 + 1] = i + 1;
```

Comment:
why is this set to `i+1` if we only want to do initialization? did we want to create bonds between `(0,1), (0,3), (0,5)`?

Reply:
(is N_ == K here?)

Reply:
What we want is for `default_bonds[i * 2 + 0] != default_bonds[i * 2 + 1]` as that is verified in the constructor. This was a way to ensure that, could do it pretty much any way we want.

Will add comment for now, and we can determine a different approach later

> (is N_ == K here?)

Yes

Reply:
Added comment here: https://github.com/proteneer/timemachine/pull/1004/commits/5621c2dc9fbd1ac77b6205c62de668ed233c6a29